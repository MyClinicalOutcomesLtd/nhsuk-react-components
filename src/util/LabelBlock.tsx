import React, { ComponentProps, ReactNode } from 'react';
import ErrorMessage from '../components/error-message/ErrorMessage';
import Hint from '../components/hint/Hint';
import Label from '../components/label/Label';

interface LabelBlockProps {
  elementID?: string;
  label?: ReactNode;
  labelProps?: ComponentProps<typeof Label>;
  hint?: ReactNode;
  hintProps?: ComponentProps<typeof Hint>;
  error?: ReactNode | boolean;
  errorProps?: ComponentProps<typeof ErrorMessage>;
}

const LabelBlock: React.FC<LabelBlockProps> = ({
  label,
  labelProps,
  hint,
  hintProps,
  error,
  errorProps,
}) => (
  <>
    {label ? <Label {...labelProps}>{label}</Label> : null}
    {hint ? <Hint {...hintProps}>{hint}</Hint> : null}
    {error && typeof error === 'string' ? (
      <ErrorMessage {...errorProps}>{error}</ErrorMessage>
    ) : null}
  </>
);

export default LabelBlock;
