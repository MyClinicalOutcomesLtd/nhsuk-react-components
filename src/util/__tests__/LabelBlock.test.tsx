import { mount } from 'enzyme';
import React from 'react';
import LabelBlock from '../LabelBlock';

describe('LabelBlock', () => {
  it.skip('renders a label element when label is provided and passes props', () => {
    const wrapper = mount(
      <LabelBlock
        elementId="test-element"
        label="Test Label"
        labelProps={{ 'aria-label': 'Test Aria Label' }}
      />,
    );

    expect(wrapper).toMatchSnapshot();

    const labelEl = wrapper.find('label.nhsuk-label');

    expect(labelEl.exists()).toBe(true);
    expect(labelEl.text()).toEqual('Test Label');
    expect(labelEl.prop('id')).toEqual('test-element--label');
    expect(labelEl.prop('aria-label')).toEqual('Test Aria Label');

    wrapper.unmount();
  });

  it.skip('renders a hint element when hint is provided and passes props', () => {
    const wrapper = mount(
      <LabelBlock
        elementId="test-element"
        hint="Test Hint"
        hintProps={{ 'aria-label': 'Test Aria Label for Hint' }}
      />,
    );

    expect(wrapper).toMatchSnapshot();

    const hintEl = wrapper.find('div.nhsuk-hint');

    expect(hintEl.exists()).toBe(true);
    expect(hintEl.text()).toEqual('Test Hint');
    expect(hintEl.prop('id')).toEqual('test-element--hint');
    expect(hintEl.prop('aria-label')).toEqual('Test Aria Label for Hint');

    wrapper.unmount();
  });

  it.skip('renders an errormessage element when error is provided and passes props', () => {
    const wrapper = mount(
      <LabelBlock
        elementId="test-element"
        error="Test Error"
        errorProps={{ 'aria-label': 'Test Aria Label for Error' }}
      />,
    );

    expect(wrapper).toMatchSnapshot();

    const errorEl = wrapper.find('span.nhsuk-error-message');

    expect(errorEl.exists()).toBe(true);
    expect(errorEl.text()).toEqual('Error: Test Error');
    expect(errorEl.prop('id')).toEqual('test-element--error-message');
    expect(errorEl.prop('aria-label')).toEqual('Test Aria Label for Error');

    wrapper.unmount();
  });

  it('does not render an errormessage element when error is provided but is boolean', () => {
    const wrapper = mount(<LabelBlock error={true} />);

    expect(wrapper).toMatchSnapshot();

    const errorEl = wrapper.find('.nhsuk-error-message');
    expect(errorEl.exists()).toBe(false);

    wrapper.unmount();
  });

  it.skip('does not add elementId to elements if no ID is provided', () => {
    const wrapper = mount(<LabelBlock label="Test Label" hint="Test Hint" error="Test Error" />);

    const labelEl = wrapper.find('label.nhsuk-label');
    const hintEl = wrapper.find('div.nhsuk-hint');
    const errorEl = wrapper.find('span.nhsuk-error-message');

    expect(labelEl.prop('id')).toBe(undefined);
    expect(hintEl.prop('id')).toBe(undefined);
    expect(errorEl.prop('id')).toBe(undefined);
  });

  it('renders nothing when no details passed', () => {
    const wrapper = mount(<LabelBlock elementId="test-id" />);
    expect(wrapper.isEmptyRender()).toBe(true);
    wrapper.unmount();
  });
});
