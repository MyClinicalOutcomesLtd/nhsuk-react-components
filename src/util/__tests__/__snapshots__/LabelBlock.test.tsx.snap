// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelBlock does not render an errormessage element when error is provided but is boolean 1`] = `
<LabelBlock
  error={true}
/>
`;

exports[`LabelBlock renders a hint element when hint is provided and passes props 1`] = `
<LabelBlock
  elementId="test-element"
  hint="Test Hint"
  hintProps={
    Object {
      "aria-label": "Test Aria Label for Hint",
    }
  }
>
  <Hint
    aria-label="Test Aria Label for Hint"
  >
    <div
      aria-label="Test Aria Label for Hint"
      className="nhsuk-hint"
    >
      Test Hint
    </div>
  </Hint>
</LabelBlock>
`;

exports[`LabelBlock renders a label element when label is provided and passes props 1`] = `
<LabelBlock
  elementId="test-element"
  label="Test Label"
  labelProps={
    Object {
      "aria-label": "Test Aria Label",
    }
  }
>
  <Label
    aria-label="Test Aria Label"
  >
    <BaseLabel
      aria-label="Test Aria Label"
    >
      <label
        aria-label="Test Aria Label"
        className="nhsuk-label"
      >
        Test Label
      </label>
    </BaseLabel>
  </Label>
</LabelBlock>
`;

exports[`LabelBlock renders an errormessage element when error is provided and passes props 1`] = `
<LabelBlock
  elementId="test-element"
  error="Test Error"
  errorProps={
    Object {
      "aria-label": "Test Aria Label for Error",
    }
  }
>
  <ErrorMessage
    aria-label="Test Aria Label for Error"
    role="alert"
    visuallyHiddenText="Error: "
  >
    <span
      aria-label="Test Aria Label for Error"
      className="nhsuk-error-message"
      role="alert"
    >
      <span
        className="nhsuk-u-visually-hidden"
      >
        Error: 
      </span>
      Test Error
    </span>
  </ErrorMessage>
</LabelBlock>
`;
